# Proyecto

### 1. Importación de librerías y archivos

In [1]:
# Importar las librerías necesarias según el análisis que se vaya a realizar
# Librería para manejo de datos convencional
import pandas as pd

In [2]:
# cargar los datos en csv
data= pd.read_csv('empleadosTecnologiaAlpes.csv', sep=';')
# Visualizar los datos
data.head()

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
0    41       Si     Viaja-Pocas-Veces            1102   
1    49       No  Viaja-Frecuentemente             279   
2    37       Si     Viaja-Pocas-Veces            1373   
3    33       No  Viaja-Frecuentemente            1392   
4    27       No     Viaja-Pocas-Veces             591   

                 Departamento  DistanciaDesdeCasa  Educacion  \
0                      Ventas                   1          2   
1  Investigacion & Desarrollo                   8          1   
2  Investigacion & Desarrollo                   2          2   
3  Investigacion & Desarrollo                   3          4   
4  Investigacion & Desarrollo                   2          1   

        CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  HorasEstandar  \
0  Ciencias de la vida              1.0               1  ...             80   
1  Ciencias de la vida              1.0               2  ...             80   
2                 Otro              1.0               4  ...             80   
3  Ciencias de la vida              1.0               5  ...             80   
4             Medicina              1.0               7  ...             80   

  PlanSalarioComplementario  TotalAnniosTrabajados  FormacionUltimoAnnio  \
0                         0                      8                     0   
1                         1                     10                     3   
2                         0                      7                     3   
3                         0                      8                     3   
4                         1                      6                     3   

   BalanceTrabajoVida FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
0                   1          26/07/2015                6               4   
1                   3          15/02/2011               10               7   
2                   3           2/06/2021                0               0   
3                   3          26/09/2013                8               7   
4                   3           2/03/2019                2               2   

   AnniosDesdeUltimaPromocion  AnniosActualDirector  
0                           0                     5  
1                           1                     7  
2                           0                     0  
3                           3                     0  
4                           2                     2  

[5 rows x 36 columns]

### 2. Análisis de la fuente de los datos

### 2.1. Registros y atributos

In [3]:
# Tamaño del dataset (filas, columnas)
data.shape

(1492, 36)

In [4]:
# Revisión de los atributos
data.columns

Index(['Edad', 'Retirado', 'ViajesDeNegocio', 'TarifaSalarial', 'Departamento',
       'DistanciaDesdeCasa', 'Educacion', 'CampoEducacion', 'EmpleadosACargo',
       'NumeroEmpleado', 'NivelSatisfaccion', 'Genero', 'SalarioPorHora',
       'NivelInvolucramiento', 'NivelResponsabilidad', 'Rol',
       'SatisfaccionTrabajo', 'EstadoCivil', 'SalarioMensual', 'TarifaMensual',
       'NumEmpresasDondeTrabaja', 'MasDe18Horas', 'HorasExtra',
       'PorcentajeIncrementoSalarioUltimoAnnio', 'NivelRendimiento',
       'SatisfaccionEmpresa', 'HorasEstandar', 'PlanSalarioComplementario',
       'TotalAnniosTrabajados', 'FormacionUltimoAnnio', 'BalanceTrabajoVida',
       'FechaIngresoEmpresa', 'AnniosEnEmpresa', 'AnniosRolActual',
       'AnniosDesdeUltimaPromocion', 'AnniosActualDirector'],
      dtype='object')

### 2.2. Tipos de datos

In [5]:
#identificar los tipos de datos
data.dtypes

Edad                                        int64
Retirado                                   object
ViajesDeNegocio                            object
TarifaSalarial                              int64
Departamento                               object
DistanciaDesdeCasa                          int64
Educacion                                   int64
CampoEducacion                             object
EmpleadosACargo                           float64
NumeroEmpleado                              int64
NivelSatisfaccion                           int64
Genero                                     object
SalarioPorHora                              int64
NivelInvolucramiento                        int64
NivelResponsabilidad                        int64
Rol                                        object
SatisfaccionTrabajo                         int64
EstadoCivil                                object
SalarioMensual                              int64
TarifaMensual                               int64


### 3. Completitud
### 3.1. Identificar los atributos vacíos

In [6]:
# conteo de los datos vacíos por atributo
data.isna().sum()

Edad                                       0
Retirado                                   0
ViajesDeNegocio                            0
TarifaSalarial                             0
Departamento                               0
DistanciaDesdeCasa                         0
Educacion                                  0
CampoEducacion                             0
EmpleadosACargo                           20
NumeroEmpleado                             0
NivelSatisfaccion                          0
Genero                                     7
SalarioPorHora                             0
NivelInvolucramiento                       0
NivelResponsabilidad                       0
Rol                                        0
SatisfaccionTrabajo                        0
EstadoCivil                               10
SalarioMensual                             0
TarifaMensual                              0
NumEmpresasDondeTrabaja                    0
MasDe18Horas                               0
HorasExtra

Los atributos empleadosACargo y estadoCivil presentan datos vacios.

In [7]:
# porcentaje de los datos vacíos por atributo
data[['EmpleadosACargo','EstadoCivil']].isna().sum()/len(data)

EmpleadosACargo    0.013405
EstadoCivil        0.006702
dtype: float64

### 4. Análisis de datos duplicados

### 4.1. Identificar el identificador único

In [8]:
### Contar los identificadores únicos
len(data['NumeroEmpleado'].unique())

1470

El número de registros de nuestra fuente es de 1492 registros. Sin embargo, existen únicamente 1470 identificadores únicos. Un  claro indicador de que existen registros duplicados.

### 4.2. Identificar los valores duplicados

Debemos identificar dónde se encuentran los datos duplicados, podemos ver si en particular es en un solo atributo, o todos los atributos se encuentran duplicados, en las siguientes líneas vamos a desarrollar los dos análisis.

In [9]:
# revisar si el problema de duplicados está en un solo atributo, en este caso el número de orden
data[data.duplicated('NumeroEmpleado')]

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
1201    23       Si     Viaja-Pocas-Veces             427   
1438    23       Si  Viaja-Frecuentemente             638   
1470    26       Si     Viaja-Pocas-Veces            1357   
1471    20       Si  Viaja-Frecuentemente             871   
1472    26       Si     Viaja-Pocas-Veces            1449   
1473    26       Si     Viaja-Pocas-Veces             950   
1474    26       Si  Viaja-Frecuentemente             575   
1475    26       Si  Viaja-Frecuentemente             426   
1476    26       Si     Viaja-Pocas-Veces             471   
1477    20       Si     Viaja-Pocas-Veces            1362   
1478    26       Si     Viaja-Pocas-Veces            1146   
1480    26       Si  Viaja-Frecuentemente             887   
1481    20       Si     Viaja-Pocas-Veces             500   
1482    20       Si     Viaja-Pocas-Veces             129   
1483    20       Si     Viaja-Pocas-Veces            1097   
1484    26       Si              No-Viaja             265   
1485    26       Si  Viaja-Frecuentemente             342   
1486    20       Si  Viaja-Frecuentemente             769   
1487    26       Si     Viaja-Pocas-Veces            1330   
1488    23       Si     Viaja-Pocas-Veces            1243   
1490    26       Si     Viaja-Pocas-Veces             920   
1491    23       Si     Viaja-Pocas-Veces            1320   

                    Departamento  DistanciaDesdeCasa  Educacion  \
1201                      Ventas                   7          3   
1438                      Ventas                   9          3   
1470  Investigacion & Desarrollo                  25          3   
1471  Investigacion & Desarrollo                   6          3   
1472  Investigacion & Desarrollo                  16          4   
1473                      Ventas                   4          4   
1474  Investigacion & Desarrollo                   3          1   
1475              Talento Humano                  17          4   
1476  Investigacion & Desarrollo                  24          3   
1477  Investigacion & Desarrollo                  10          1   
1478                      Ventas                   8          3   
1480  Investigacion & Desarrollo                   5          2   
1481                      Ventas                   2          3   
1482  Investigacion & Desarrollo                   4          3   
1483  Investigacion & Desarrollo                  11          3   
1484                      Ventas                  29          2   
1485  Investigacion & Desarrollo                   2          3   
1486                      Ventas                   9          3   
1487  Investigacion & Desarrollo                  21          3   
1488  Investigacion & Desarrollo                   6          3   
1490              Talento Humano                  20          2   
1491  Investigacion & Desarrollo                   8          1   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
1201  Ciencias de la vida              1.0            1702  ...   
1438             Medicina              1.0            2023  ...   
1470  Ciencias de la vida              1.0              55  ...   
1471  Ciencias de la vida              1.0             137  ...   
1472             Medicina              1.0             394  ...   
1473             Mercadeo              1.0             401  ...   
1474        Nivel tecnico              1.0             510  ...   
1475  Ciencias de la vida              1.0             608  ...   
1476        Nivel tecnico              1.0             622  ...   
1477             Medicina              1.0             701  ...   
1478        Nivel tecnico              1.0             796  ...   
1480             Medicina              1.0             848  ...   
1481             Medicina              1.0             922  ...   
1482        Nivel tecnico              1.0             960  ...   
1483             Medicina              1.0            1

In [10]:
# revisar si hay registros donde todos los atributos están duplicados (son iguales)
data[data.duplicated()]

Edad Retirado       ViajesDeNegocio  TarifaSalarial  \
1201    23       Si     Viaja-Pocas-Veces             427   
1470    26       Si     Viaja-Pocas-Veces            1357   
1471    20       Si  Viaja-Frecuentemente             871   
1472    26       Si     Viaja-Pocas-Veces            1449   
1473    26       Si     Viaja-Pocas-Veces             950   
1474    26       Si  Viaja-Frecuentemente             575   
1475    26       Si  Viaja-Frecuentemente             426   
1476    26       Si     Viaja-Pocas-Veces             471   
1477    20       Si     Viaja-Pocas-Veces            1362   
1478    26       Si     Viaja-Pocas-Veces            1146   
1480    26       Si  Viaja-Frecuentemente             887   
1481    20       Si     Viaja-Pocas-Veces             500   
1482    20       Si     Viaja-Pocas-Veces             129   
1483    20       Si     Viaja-Pocas-Veces            1097   
1484    26       Si              No-Viaja             265   
1485    26       Si  Viaja-Frecuentemente             342   
1486    20       Si  Viaja-Frecuentemente             769   
1487    26       Si     Viaja-Pocas-Veces            1330   
1488    23       Si     Viaja-Pocas-Veces            1243   
1490    26       Si     Viaja-Pocas-Veces             920   
1491    23       Si     Viaja-Pocas-Veces            1320   

                    Departamento  DistanciaDesdeCasa  Educacion  \
1201                      Ventas                   7          3   
1470  Investigacion & Desarrollo                  25          3   
1471  Investigacion & Desarrollo                   6          3   
1472  Investigacion & Desarrollo                  16          4   
1473                      Ventas                   4          4   
1474  Investigacion & Desarrollo                   3          1   
1475              Talento Humano                  17          4   
1476  Investigacion & Desarrollo                  24          3   
1477  Investigacion & Desarrollo                  10          1   
1478                      Ventas                   8          3   
1480  Investigacion & Desarrollo                   5          2   
1481                      Ventas                   2          3   
1482  Investigacion & Desarrollo                   4          3   
1483  Investigacion & Desarrollo                  11          3   
1484                      Ventas                  29          2   
1485  Investigacion & Desarrollo                   2          3   
1486                      Ventas                   9          3   
1487  Investigacion & Desarrollo                  21          3   
1488  Investigacion & Desarrollo                   6          3   
1490              Talento Humano                  20          2   
1491  Investigacion & Desarrollo                   8          1   

           CampoEducacion  EmpleadosACargo  NumeroEmpleado  ...  \
1201  Ciencias de la vida              1.0            1702  ...   
1470  Ciencias de la vida              1.0              55  ...   
1471  Ciencias de la vida              1.0             137  ...   
1472             Medicina              1.0             394  ...   
1473             Mercadeo              1.0             401  ...   
1474        Nivel tecnico              1.0             510  ...   
1475  Ciencias de la vida              1.0             608  ...   
1476        Nivel tecnico              1.0             622  ...   
1477             Medicina              1.0             701  ...   
1478        Nivel tecnico              1.0             796  ...   
1480             Medicina              1.0             848  ...   
1481             Medicina              1.0             922  ...   
1482        Nivel tecnico              1.0             960  ...   
1483             Medicina              1.0            1016  ...   
1484             Medicina              1.0            1037  ...   
1485  Ciencias de la vida              1.0            1053  ...   
1486             Mercadeo              1.0       

Como vemos, existen 21 datos que se repiten en todos los atributos y también está el caso de 1 registros que se repiten en el número de orden, pero no necesariamente es igual en los otros atributos. Vamos a revisar estos casos en particular para entenderlos mejor.

In [11]:
# revisar los registros que tienen el número de orden duplicado pero no son iguales los valores de los dempas atributos
data[data.duplicated('NumeroEmpleado') & ~data.duplicated()]

Edad Retirado       ViajesDeNegocio  TarifaSalarial Departamento  \
1438    23       Si  Viaja-Frecuentemente             638       Ventas   

      DistanciaDesdeCasa  Educacion CampoEducacion  EmpleadosACargo  \
1438                   9          3       Medicina              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
1438            2023  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
1438                      1                     3                   2   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
1438          25/04/2020                1               0   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
1438                           1                     0  

[1 rows x 36 columns]

In [12]:
data[data['NumeroEmpleado']==2023]

Edad Retirado       ViajesDeNegocio  TarifaSalarial Departamento  \
1213    23       Si  Viaja-Frecuentemente             638       Ventas   
1438    23       Si  Viaja-Frecuentemente             638       Ventas   

      DistanciaDesdeCasa  Educacion CampoEducacion  EmpleadosACargo  \
1213                   9          3       Mercadeo              1.0   
1438                   9          3       Medicina              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
1213            2023  ...             80                         1   
1438            2023  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
1213                      1                     3                   2   
1438                      1                     3                   2   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
1213          25/04/2020                1               0   
1438          25/04/2020                1               0   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
1213                           1                     0  
1438                           1                     0  

[2 rows x 36 columns]

In [13]:
# revisar en particular las órdenes que tiene este comportamiento
numero_ordenes_duplicadas=data[data.duplicated('NumeroEmpleado') & ~data.duplicated()]['NumeroEmpleado']
#numero_ordenes_duplicadas
data[data['NumeroEmpleado'].isin(numero_ordenes_duplicadas)]

Edad Retirado       ViajesDeNegocio  TarifaSalarial Departamento  \
1213    23       Si  Viaja-Frecuentemente             638       Ventas   
1438    23       Si  Viaja-Frecuentemente             638       Ventas   

      DistanciaDesdeCasa  Educacion CampoEducacion  EmpleadosACargo  \
1213                   9          3       Mercadeo              1.0   
1438                   9          3       Medicina              1.0   

      NumeroEmpleado  ...  HorasEstandar PlanSalarioComplementario  \
1213            2023  ...             80                         1   
1438            2023  ...             80                         1   

      TotalAnniosTrabajados  FormacionUltimoAnnio  BalanceTrabajoVida  \
1213                      1                     3                   2   
1438                      1                     3                   2   

     FechaIngresoEmpresa  AnniosEnEmpresa AnniosRolActual  \
1213          25/04/2020                1               0   
1438          25/04/2020                1               0   

      AnniosDesdeUltimaPromocion  AnniosActualDirector  
1213                           1                     0  
1438                           1                     0  

[2 rows x 36 columns]

Vemos que los registros con índices:
+ 1213 y 1438

Son exactamente los mismos registros, solamente que el campoEducacion en un registro es Mercadeo y en el otro Medecina. 

### 4.3. Eliminación de duplicados

In [14]:
#Eliminar los registros de columnas que están con la misma información
#data=data.drop([1438])
#data.head()

In [15]:
# eliminar el resto de duplicados
#data=data.drop_duplicates()
#data.head()

In [16]:
#validar que no existan duplicados
# longitud de la base
print(data.shape)
print(data.duplicated().sum())

(1492, 36)
21


Como vemos, después de eliminar los datos duplicados, se obtienen 1150 registros, que son los que desde un inicio deberíamos dejar como resultado para utilizar en análisis futuros.

### 5. Estandarización

### 5.1. Estandarizar un atributo tipo fecha

In [17]:
# revisar el formato del atributo  fecha_envío_límite
data[['FechaIngresoEmpresa']].head(30)

FechaIngresoEmpresa
0           26/07/2015
1           15/02/2011
2            2/06/2021
3           26/09/2013
4            2/03/2019
5            8/07/2014
6            6/07/2020
7           15/08/2020
8           13/10/2012
9           15/07/2014
10          26/05/2016
11          11/07/2012
12           9/02/2016
13          15/03/2019
14          19/09/2017
15           2/02/2011
16           6/04/2015
17           1/05/2020
18           7/01/1996
19          14/11/2018
20           8/04/2017
21          15/05/2016
22           4/01/2009
23          23/04/2021
24          21/08/2017
25          16/02/2007
26           3/03/2011
27           3/09/2012
28          15/01/1999
29          26/08/2019

La fecha no presenta problemas en el formato, toda la columna posee el mismo formato.

In [18]:
# convertir a datetime el atributo FechaIngresoEmpresa
data['FechaIngresoEmpresa']=pd.to_datetime(data['FechaIngresoEmpresa'])
data[['FechaIngresoEmpresa']].head()

C:\Users\Edwar\AppData\Local\Temp\ipykernel_2596\1988937927.py:2: UserWarning: Parsing '26/07/2015' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  data['FechaIngresoEmpresa']=pd.to_datetime(data['FechaIngresoEmpresa'])
C:\Users\Edwar\AppData\Local\Temp\ipykernel_2596\1988937927.py:2: UserWarning: Parsing '15/02/2011' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  data['FechaIngresoEmpresa']=pd.to_datetime(data['FechaIngresoEmpresa'])
C:\Users\Edwar\AppData\Local\Temp\ipykernel_2596\1988937927.py:2: UserWarning: Parsing '26/09/2013' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  data['FechaIngresoEmpresa']=pd.to_datetime(data['FechaIngresoEmpresa'])
C:\Users\Edwar\AppData\Local\Temp\ipykernel_2596\1988937927.py:2: UserWarning: Parsing '15/08/2020' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for co

FechaIngresoEmpresa
0          2015-07-26
1          2011-02-15
2          2021-02-06
3          2013-09-26
4          2019-02-03

In [19]:
# revisar que el tipo de dato haya quedado correcto
data[['FechaIngresoEmpresa']].dtypes

FechaIngresoEmpresa    datetime64[ns]
dtype: object

### 5.3. Estandarizar un atributo categórico

In [20]:
#identificar las columnas de tipo string, si no se toma de la descripción inicial de la fuente de datos
data.select_dtypes(include='object').columns

Index(['Retirado', 'ViajesDeNegocio', 'Departamento', 'CampoEducacion',
       'Genero', 'Rol', 'EstadoCivil', 'MasDe18Horas', 'HorasExtra'],
      dtype='object')

In [21]:
# Consultar los datos del atributo Retirado
data['Retirado'].value_counts()

No    1233
Si     259
Name: Retirado, dtype: int64

No presenta problema.

In [22]:
# Consultar los datos del atributo ViajesDeNegocio
data['ViajesDeNegocio'].value_counts()

Viaja-Pocas-Veces       1057
Viaja-Frecuentemente     284
No-Viaja                 151
Name: ViajesDeNegocio, dtype: int64

No presenta problema.

In [23]:
# Consultar los datos del atributo Departamento
data['Departamento'].value_counts()

Investigacion & Desarrollo    974
Ventas                        451
Talento Humano                 65
Venta                           1
Venta                           1
Name: Departamento, dtype: int64

Presenta problema.

In [24]:
# Consultar los datos del atributo CampoEducacion
data['CampoEducacion'].value_counts()

Ciencias de la vida    612
Medicina               471
Mercadeo               159
Nivel tecnico          136
Otro                    82
Talento Humano          27
medicina                 2
mercaeo                  2
meDicina                 1
Name: CampoEducacion, dtype: int64

Presenta problema.

In [25]:
# Consultar los datos del atributo Genero
data['Genero'].value_counts()

Masculino    896
Femenino     589
Name: Genero, dtype: int64

No presenta problema.

In [26]:
# Consultar los datos del atributo Rol
data['Rol'].value_counts()

Ejecutivo de ventas              329
Cientifico de investigacion      297
Tecnico de laboratorio           267
Director de manufactura          145
Representante atencion medica    131
Director                         102
Representante de ventas           87
Director de investigacion         80
Talento Humano                    54
Name: Rol, dtype: int64

No presenta problema.

In [27]:
# Consultar los datos del atributo EstadoCivil
data['EstadoCivil'].value_counts()

Casado        672
Soltero       482
Divorciado    328
Name: EstadoCivil, dtype: int64

No presenta problema.

In [28]:
# Consultar los datos del atributo MasDe18Horas
data['MasDe18Horas'].value_counts()

S    1492
Name: MasDe18Horas, dtype: int64

No presenta problema.

In [29]:
# Consultar los datos del atributo HorasExtra
data['HorasExtra'].value_counts()

No    1062
Si     430
Name: HorasExtra, dtype: int64

No presenta problema.

Los atriutos CampoEducacion y Departamento presentan problemas de estandarizacion.

In [30]:
data['CampoEducacion'].value_counts()

Ciencias de la vida    612
Medicina               471
Mercadeo               159
Nivel tecnico          136
Otro                    82
Talento Humano          27
medicina                 2
mercaeo                  2
meDicina                 1
Name: CampoEducacion, dtype: int64

In [31]:
data['CampoEducacion']= data['CampoEducacion'].apply(lambda x: x.capitalize())
data['CampoEducacion'].value_counts()

Ciencias de la vida    612
Medicina               474
Mercadeo               159
Nivel tecnico          136
Otro                    82
Talento humano          27
Mercaeo                  2
Name: CampoEducacion, dtype: int64

El atributo CampoEducacion quedo corregido.

In [32]:
data['Departamento'].value_counts()

Investigacion & Desarrollo    974
Ventas                        451
Talento Humano                 65
Venta                           1
Venta                           1
Name: Departamento, dtype: int64

In [33]:
def cambio(x):
    xnew=''
    if x=='Venta' or x=='Venta ':
        xnew='Ventas'
    else:
        xnew=x
    return xnew

In [34]:
data['Departamento'] = data['Departamento'].apply(cambio)
data['Departamento'].value_counts()

Investigacion & Desarrollo    974
Ventas                        453
Talento Humano                 65
Name: Departamento, dtype: int64

El atributo Departamento quedo corregido.